### ISSUE TITLE
TensorFlow Lite Android nightly method "runForMultipleInputsOutputs" that loaded model using stft  not work with converted by tf.lite.OpsSet.SELECT_TF_OPS 

### ISSUE THREAD

**Issue by tk-okumura** [tensorflow/tensorflow#42723]
TensorFlow Lite Android nightly method "runForMultipleInputsOutputs" that loaded model using stft  not work with converted by tf.lite.OpsSet.SELECT_TF_OPS .
What is wrong ?

written code.
> import numpy as np
> import tensorflow as tf
> from tensorflow.keras.layers import Input, Dense, Dropout
> from tensorflow.keras.layers import Conv2D, MaxPool2D,Flatten
> from tensorflow.keras.layers import BatchNormalization,Activation,Add,GlobalAveragePooling2D,Reshape
> from tensorflow.keras.models import Model
> 
> _FFT_SIZE=512
> _HOP_SIZE=256
> _N_MEL_BINS=128
> 
> class LogMelSpectrogram(tf.keras.layers.Layer):
>     """Compute log-magnitude mel-scaled spectrograms."""
> 
>     def __init__(self, sample_rate, fft_size, hop_size, n_mels,
>                  f_min=0.0, f_max=None, **kwargs):
>         super(LogMelSpectrogram, self).__init__(**kwargs)
>         self.sample_rate = sample_rate
>         self.fft_size = fft_size
>         self.hop_size = hop_size
>         self.n_mels = n_mels
>         self.f_min = f_min
>         self.f_max = f_max if f_max else sample_rate / 2
>         self.mel_filterbank = tf.signal.linear_to_mel_weight_matrix(
>             num_mel_bins=self.n_mels,
>             num_spectrogram_bins=fft_size // 2 + 1,
>             sample_rate=self.sample_rate,
>             lower_edge_hertz=self.f_min,
>             upper_edge_hertz=self.f_max)
> 
>     def build(self, input_shape):
>         self.non_trainable_weights.append(self.mel_filterbank)
>         super(LogMelSpectrogram, self).build(input_shape)
> 
>     def call(self, waveforms):
>         """Forward pass.
>         Parameters
>         ----------
>         waveforms : tf.Tensor, shape = (None, n_samples)
>             A Batch of mono waveforms.
>         Returns
>         -------
>         log_mel_spectrograms : (tf.Tensor), shape = (None, time, freq, ch)
>             The corresponding batch of log-mel-spectrograms
>         """
>         def _tf_log10(x):
>             numerator = tf.math.log(x)
>             denominator = tf.math.log(tf.constant(10, dtype=numerator.dtype))
>             return numerator / denominator
> 
>         def power_to_db(magnitude, amin=1e-16, top_db=80.0):
>             """
>             https://librosa.github.io/librosa/generated/librosa.core.power_to_db.html
>             """
>             ref_value = tf.reduce_max(magnitude)
>             log_spec = 10.0 * _tf_log10(tf.maximum(amin, magnitude))
>             log_spec -= 10.0 * _tf_log10(tf.maximum(amin, ref_value))
>             log_spec = tf.maximum(log_spec, tf.reduce_max(log_spec) - top_db)
> 
>             return log_spec
> 
>         spectrograms = tf.signal.stft(waveforms,
>                                       frame_length=self.fft_size,
>                                       frame_step=self.hop_size,
>                                       pad_end=False)
> 
>         magnitude_spectrograms = tf.abs(spectrograms)
> 
>         mel_spectrograms = tf.matmul(tf.square(magnitude_spectrograms),
>                                      self.mel_filterbank)
> 
>         log_mel_spectrograms = power_to_db(mel_spectrograms)
> 
>         # add channel dimension
>         log_mel_spectrograms = tf.expand_dims(log_mel_spectrograms, 3)
> 
>         return log_mel_spectrograms
> 
>     def get_config(self):
>         config = {
>             'fft_size': self.fft_size,
>             'hop_size': self.hop_size,
>             'n_mels': self.n_mels,
>             'sample_rate': self.sample_rate,
>             'f_min': self.f_min,
>             'f_max': self.f_max,
>         }
>         config.update(super(LogMelSpectrogram, self).get_config())
> 
>         return config
> 
> 
> 
> def GetModel(n_classes, sample_rate=16000, duration=4,
>               fft_size=_FFT_SIZE, hop_size=_HOP_SIZE, n_mels=_N_MEL_BINS):
>     n_samples = sample_rate * duration
>     input_data = Input(shape=(n_samples,), name='input', dtype='float32')
>     freq = Input(name='freq',shape=[1],dtype='int32')
>     x = input_data
>     test_output = LogMelSpectrogram(sample_rate, fft_size, hop_size, n_mels)(x)
>     x = Dense(n_classes,activation='softmax')(x)
> 
>     return Model(inputs=[input_data,freq], outputs=x)
> 

convert model to tflite with tf.lite.OpsSet.SELECT_TF_OPS
> N_classes = 50
> Duration = 5
> 
> model = GetModel(N_classes,duration=Duration)
> opt = tf.keras.optimizers.Adam(lr=0.001, decay=1e-6, amsgrad=True)
> model.compile(loss='categorical_crossentropy',
>           optimizer=opt,
>           metrics=['accuracy'])
> model.summary()
> 
> 
> converter = tf.lite.TFLiteConverter.from_keras_model(model)
> 
> #converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS]
> #converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS]
> converter.target_spec.supported_ops = [tf.lite.OpsSet.SELECT_TF_OPS]
> 
> tflite_model = converter.convert()
> with open("model.tflite", "wb") as f:
>     f.write(tflite_model)

on Centos , tflite invoke from python is work.
>   input_data = np.zeros(input_shape, dtype=np.float32)
>   pprint("input_data")
>   pprint(input_data)
>   input_data[0][0:] = x2
>   npa = np.array([[0]], dtype=np.int32) 
>   npa[0] = 16000
> 
>   print(f"input_details[0]:{input_details[0]}")
>   print(f"input_details[1]:{input_details[1]}")
> 
>   interpreter.set_tensor(input_details[0]['index'], input_data)
>   interpreter.set_tensor(input_details[1]['index'], npa)
> 
>   interpreter.invoke()
>   output_data = interpreter.get_tensor(output_details[0]['index'])
>   print("output_data:")
>   pprint(output_data)
> 

Results:
> array([[0.01259949, 0.01274813, 0.02341373, 0.02667766, 0.01372223,
>         0.02063654, 0.01939304, 0.01231323, 0.02360482, 0.01985866,
>         0.01616379, 0.02790864, 0.01614609, 0.03300844, 0.02038008,
>         0.02689514, 0.01205852, 0.02373471, 0.03396689, 0.0159423 ,
>         0.01418541, 0.01825977, 0.01506583, 0.01811898, 0.0167273 ,
>         0.01219947, 0.03590814, 0.01607339, 0.0213427 , 0.02323929,
>         0.01976353, 0.01719474, 0.02329947, 0.02231019, 0.02448872,
>         0.01822161, 0.01806694, 0.02725129, 0.0257786 , 0.02799631,
>         0.02095988, 0.01663861, 0.01729359, 0.0102047 , 0.01404527,
>         0.01819064, 0.0125629 , 0.01387509, 0.01704353, 0.032522  ]],
>       dtype=float32)
>

on Android , Kotlin
>tfLite!!.runForMultipleInputsOutputs(inputArray, outputs) 

example 1
>Internal error: Failed to run on the given Interpreter: Matrix size-incompatible: In[0]: [80000,1], In[1]: [80000,50]
>    	 (while executing 'MatMul' via Eager)

example 2
>Cannot copy from a TensorFlowLite tensor (Identity) with shape [80000, 50] to a Java object with shape [1, 50].

**Comment by Saduf2019** [comment]
@tk-okumura 
Can you please provide indented complete code, or share a colab gist with the error reported.

**Comment by tk-okumura** [comment]
Hi Saduf2019.
thanks your reply.

after to post issue, I tryed to write minimum code.

I needed to use stft.
It need to convert with "tf.lite.OpsSet.SELECT_TF_OPS".
and so converted model rise error.

**error (written by kotlin)**

>    I/System.out: java.lang.IllegalArgumentException: Internal error: Failed to run on the given Interpreter: Matrix size-incompatible: 
>   In[0]: [80000,1], In[1]: [80000,50]
>        	 (while executing 'MatMul' via Eager)
>        Node number 6 (TfLiteFlexDelegate) failed to invoke.
> 

Python code "check for runnable" isn't return any error.
I comment out OPS option "(*1)", Kotlin isn't return any error.
(test code is not contain stft then can convert without SELECT_TF_OPS.)

Kotlin code
build.gradle

>     implementation 'org.tensorflow:tensorflow-lite:0.0.0-nightly'
>     implementation 'org.tensorflow:tensorflow-lite-select-tf-ops:0.0.0-nightly'
> 

Python
Model

>   import numpy as np
>   import tensorflow as tf
>   from tensorflow.keras.layers import Input, Dense, Dropout
>   from tensorflow.keras.layers import Conv2D, MaxPool2D,Flatten
>   from tensorflow.keras.layers import BatchNormalization,Activation,Add,GlobalAveragePooling2D,Reshape
>   from tensorflow.keras.models import Model
>   
>   _FFT_SIZE=512
>   _HOP_SIZE=256
>   _N_MEL_BINS=128
>   
>   def GetModel(n_classes, sample_rate=16000, duration=4,
>                 fft_size=_FFT_SIZE, hop_size=_HOP_SIZE, n_mels=_N_MEL_BINS):
>       n_samples = sample_rate * duration
>       input_data = Input(shape=(n_samples,), name='input', dtype='float32')
>       x = input_data
>       x = Dense(n_classes,activation='softmax')(x)
>       return Model(inputs=input_data, outputs=x)
> 
> Convert
>   import os
>   import tensorflow as tf
>   from declareReportSimple import GetModel
>   
>   N_classes = 50
>   Duration = 5
>   
>   model = GetModel(N_classes,duration=Duration)
>   #opt = tf.keras.optimizers.Adam(lr=0.001, decay=1e-6, amsgrad=True)
>   opt = tf.keras.optimizers.Adam()
>   model.compile(loss='categorical_crossentropy',
>             optimizer=opt,
>             metrics=['accuracy'])
>   model.summary()
>   
>   converter = tf.lite.TFLiteConverter.from_keras_model(model)
>   
>   #converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS]
>   #converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS]
>   converter.target_spec.supported_ops = [tf.lite.OpsSet.SELECT_TF_OPS] # (*1)
>   
>   tflite_model = converter.convert()
>   with open("for_report_simple.tflite", "wb") as f:
>       f.write(tflite_model)
> 
> 

check for runnable

>   import os
>   import pandas as pd
>   import numpy as np
>   import tensorflow as tf
>   from tensorflow.keras.utils import to_categorical
>   from pprint import pprint
>   
>   cur_dir = os.getcwd()
>   tflitemodel = 'for_report_simple.tflite'
>   
>   N_classes = 50
>   Duration = 5
>   
>   interpreter = tf.lite.Interpreter(model_path=cur_dir+'/'+tflitemodel)
>   interpreter.allocate_tensors()
>   input_details = interpreter.get_input_details()
>   output_details = interpreter.get_output_details()
>   
>   input_shape = input_details[0]['shape']
>   
>   pprint(f"input_details[0]:{input_details[0]}")
>   pprint(f"output_details[0]:{output_details[0]}")
>   
>   x2 = np.zeros((16000 * Duration))
>   
>   input_data = np.zeros(input_shape, dtype=np.float32)
>   input_data[0][0:] = x2
>   
>   pprint(f"input_data.shape:{input_data.shape}")
>   interpreter.set_tensor(input_details[0]['index'], input_data)
>   
>   interpreter.invoke()
>   output_data = interpreter.get_tensor(output_details[0]['index'])
>   print(f"output_data:{output_data}")

**Comment by tk-okumura** [comment]
resolved.

in case not use tf.lite.OpsSet.SELECT_TF_OPS,
input buffer shape [16000,1] has been processed.

in case use tf.lite.OpsSet.SELECT_TF_OPS,
input buffer shape [16000,1] has been processed  ,batch size was16000.
I changed input shape  to [1,16000]  from [16000,1].

thank you.

**Comment by ymodak** [comment]
Thanks for the update. Closing since its resolved.
